File: android/src/main/java/expo/modules/drafts/ExpoDraftsModule.kt
package expo.modules.drafts

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.functions.Queues

class ExpoDraftsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoDrafts")

    AsyncFunction("open") {
      DraftsController.open(appContext.throwingActivity)
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("setVisible") { visible: Boolean ->
      DraftsController.setVisible(appContext.throwingActivity, visible)
    }.runOnQueue(Queues.MAIN)

    Function("getState") {
      DraftsController.getState(appContext.throwingActivity)
    }
  }
}

File: ios/ExpoDraftsModule.swift
import ExpoModulesCore
import UIKit

public class ExpoDraftsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoDrafts")

    AsyncFunction("open") {
      ExpoDraftsManager.shared.open()
    }.runOnQueue(.main)

    AsyncFunction("setVisible") { (visible: Bool) in
      ExpoDraftsManager.shared.setVisible(visible)
    }.runOnQueue(.main)

    Function("getState") {
      ExpoDraftsManager.shared.state()
    }
  }
}

public class ExpoDraftsAppDelegateSubscriber: ExpoAppDelegateSubscriber {
  public func subscriberDidRegister() {
    if ExpoDraftsManager.shared.enabled {
      DraftsUpdateTransaction.recoverPendingSelection()
      ExpoDraftsManager.shared.observeUpdatesStartup()
    }
  }

  public func applicationDidBecomeActive(_ application: UIApplication) {
    ExpoDraftsManager.shared.install()
  }

  public func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    DispatchQueue.main.async {
      ExpoDraftsManager.shared.install()
    }
    return false
  }
}
